### ISSUE TITLE
new lint: "Avoid unnecessary type annotations"

### ISSUE THREAD

**Issue by pq** [dart-lang/linter#2591]
Avoid type explicit type annotations when they can be inferred.

Examples:

**GOOD:**
```dart
var strings = ['foo', 'bar', 'baz'];

var fakeSimStore = FakeStore<Sim>(Sim());

var stars = Row(
  mainAxisSize: MainAxisSize.min,
  children: [
    Icon(Icons.star, color: Colors.green[500]),
    Icon(Icons.star, color: Colors.black),
  ],
);
```

**BAD:**
```dart
var strings = <String>['foo', 'bar', 'baz'];

var fakeSimStore = FakeStore(Sim());

var stars = Row(
  mainAxisSize: MainAxisSize.min,
  children: <Widget>[
    Icon(Icons.star, color: Colors.green[500]),
    Icon(Icons.star, color: Colors.black),
  ],
);
```

/cc @jefflim-google

**Comment by a14n** [comment]
Shouldn't `fakeSimStore` examples be inversed?

**Comment by Hixie** [comment]
I don't have a problem with this lint existing, but I would strongly object to turning it on by default.

I think it makes intent harder to discern. For example, maybe you want to return a `List<Foo>` and it happens that `bar()` and `baz` are of type `Foo`, but that's not obvious:

```dart
   return [bar(), baz]; // what's the type of the list?
```
Contrast:
```dart
   return <Foo>[bar(), baz];
```

A similar problem exists with other types. For example:

```dart
   var foo = Bar(baz()); // what's the type argument to the Bar constructor?
```
Contrast:
```dart
   var foo = Bar<Quux>(baz());
```

I think it makes certain patterns ugly. For example, suppose you have three lists of `num`s, if any of them accidentally infer the wrong type then that one has to be explicitly stated:
```dart
  var a = [3, 1.5];
  var b = [1.5, 0];
  var c = <num>[0]; // have to say <num>[] otherwise it would be a List<int>
```
Contrast:
```dart
  var a = <num>[3, 1.5];
  var b = <num>[1.5, 0];
  var c = <num>[0];
```

**Comment by pq** [comment]
> Shouldn't fakeSimStore examples be inversed?

😬 

Ummmm.  Yeah.  Thanks!  🤦

**Comment by goderbauer** [comment]
I am with Hixie: In most of these cases I like the added clarity that those explicit types bring to the table. I am not against the existence of the lint, I hope it wouldn't be forced upon me, though :)

**Comment by davidmorgan** [comment]
Thanks everyone!

I suspect there is room for a enforceable (in google3/pedantic) lint here.

Note that in google3 we have already enforced [omit_local_variable_types](https://dart-lang.github.io/linter/lints/omit_local_variable_types.html), which is very much along these lines. What we found is that it removes a _lot_ of noise from the code. There were a very small number of _general_ objections to the lint, but I haven't had any reports of specific cases where it caused problems, and no requests to relax the lint after it was enforced.

I suspect we will also enforce [avoid_types_on_closure_parameters](https://dart-lang.github.io/linter/lints/avoid_types_on_closure_parameters.html) although IIRC that needs a few tweaks first.

The argument for both of those is that the important places where you explicitly specify types are in _api_ not _implementation_. E.g. the type of the thing you return is the method's return type, you don't need to also specify it in the implementation.

As usual for google3 we'll evaluate based on pre-existing violations and how the code would change to see if this looks right in practice.

Thanks :)

**Comment by eernstg** [comment]
Maybe you could consider a lint that flags cases where the declared type is non-trivial? In particular, it seems directly counterproductive if developers are punished for documenting the intended element type of a list literal with many elements of different types, just because the desired type argument `T` is exactly the least upper bound of those many types. Months later one of those elements gets deleted, the least upper bound is now a subtype `S` of what it was previously, and we get a type error when we try to add an object that is typable as a `T` but not as an `S`.

The notion of being trivial can be defined in many ways, of course, but for collection literals it could be: _A collection literal has a non-trivial element type if it contains elements (for maps: key/value pairs) that do not have the same static type._

This means that it's allowed (but not required) to specify the element type in the cases where relying on inference is obviously error prone.

**Comment by davidmorgan** [comment]
Thanks Erik. In general I agree there's room for weakening the lint for collection literals along the lines you suggest--need to figure out if that looks more or less confusing in the end :)

**Comment by munificent** [comment]
Does this lint rule apply to fields and top level variables as well? I'm 1000% excited about this for local variables and type arguments, but for fields and top level variables, we do support users annotating those explicitly even when the annotation is redundant.

**Comment by pq** [comment]
> Does this lint rule apply to fields and top level variables as well? 

I think it's all very much open to discussion and really, we should just define it to be as aligned w/ "canonical" style.  Maybe we could add a clarifying note in the style guide?

**Comment by munificent** [comment]
I think Effective Dart is already pretty clear. (Especially now that I've rewritten the type guidelines a couple of months ago.) It has separate rules for type arguments, locals, and other declarations:

* [DON’T write type arguments on generic invocations that are inferred.](https://dart.dev/guides/language/effective-dart/design#dont-write-type-arguments-on-generic-invocations-that-are-inferred)

* [DO type annotate fields and top-level variables if the type isn’t obvious.](https://dart.dev/guides/language/effective-dart/design#do-type-annotate-fields-and-top-level-variables-if-the-type-isnt-obvious)

* [DON’T redundantly type annotate initialized local variables.](https://dart.dev/guides/language/effective-dart/design#dont-redundantly-type-annotate-initialized-local-variables)

So, according to the guidelines, locals and type arguments should never be annotated redundantly, but fields and top levels can if you feel that the type isn't "obvious", which it's deliberately pretty vague about.

Maybe split it into two lints?

**Comment by pq** [comment]
I'm on the same page.

> So, according to the guidelines, locals and type arguments should never be annotated redundantly, but fields and top levels can if you feel that the type isn't "obvious", which it's deliberately pretty vague about.

💯 

This is exactly what I'd propose `unnecessary_type_annotations` enforce: locals and type args, skipping fields and top level variables entirely.

/fyi @jefflim-google

**Comment by munificent** [comment]
> This is exactly what I'd propose `unnecessary_type_annotations` enforce: locals and type args, skipping fields and top level variables entirely.

💯💯💯💯💯

I would maybe consider putting "local" in the lint name too (because I could see some users potentially wanting a separate lint that does the same for fields and top level variables).

**Comment by eernstg** [comment]
There seems to be a lot of overlap between `avoid_unnecessary_type_annotations` and `omit_obvious_local_variable_types` when it comes to local variable declarations, and an upcoming `omit_obvious_property_types` for other variables (except that `avoid_unnecessary_type_annotations` might only target locals).

The main difference would be that `avoid_unnecessary_type_annotations` would flag actual type parameters passed to generic function/method invocations or typed literals, and it would check expressions in general, not just the initializing expressions of local variable declarations (IIUC).

So maybe it could be renamed into `avoid_unnecessary_type_arguments`?
